### PR TITLE
Header footer api

### PR DESCRIPTION
**Pull request by tomdoughty**
## Description

Build header and footer from endpoint data API
Make endpoint an environment variable
Add a cache to prevent too many unnecessary calls

## Related Issue

https://jira.service.nhs.uk/browse/TM-1240

## Motivation and Context

Build header and footer links from API for consistency across applications

## Checklist

- [x] An [ADR](http://thinkrelevance.com/blog/2011/11/15/documenting-architecture-decisions)
- [ ] New and/or updated tests
- [x] Changes log in `CHANGELOG`

### CODE REVIEW

File: app/middleware/headerAndFooter.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    const footerLinks = footer.items.map(({ url, label }) => ({
+      URL: url,
+      label,
+    }));
+
+    // Set data to the cache
+    headerFooterCache.set('nhsHeaderFooterApi', {
+      footerLinks,
+      headerLinks,
+    }, 10000);
```
Comment (anlinhs) [inline app/middleware/headerAndFooter.js:42]: 10000 seconds maybe too long, I will investigate.

File: config/config.js [hunk 2]
```diff
@@ -16,6 +16,9 @@ module.exports = {
   cookiebot: {
     scriptUrl: process.env.COOKIEBOT_SCRIPT_URL || '//assets.nhs.uk/scripts/cookie-consent.js',
   },
+  headerFooterApi: {
+    endpoint: process.HEADER_FOOTER_API_ENDPOINT || 'https://nhsuk-live.nhswebsite-integration.nhs.uk/content-api/header-footer.json',
```
Comment (anlinhs) [inline config/config.js:20]: This should be process.**env.**HEADER_FOOTER_API_ENDPOINT right? 
Does this mean that if the env var is missing it defaults to the integration URL? If so, when running in the live environment could this potentially end up pointing to the integration endpoint?

File: app/middleware/headerAndFooter.js (hunk 1 above)
Comment (tomdoughty) [reply app/middleware/headerAndFooter.js]: Reduced to 5 minutes as per header/footer API devs conversation

File: config/config.js [hunk 3]
```diff
@@ -16,6 +16,10 @@ module.exports = {
   cookiebot: {
     scriptUrl: process.env.COOKIEBOT_SCRIPT_URL || '//assets.nhs.uk/scripts/cookie-consent.js',
   },
+  headerFooterApi: {
+    cacheTimeout: process.env.HEADER_FOOTER_API_CACHE_TIMEOUT || 300,
+    endpoint: process.env.HEADER_FOOTER_API_ENDPOINT || 'https://nhsuk-live.nhswebsite-integration.nhs.uk/content-api/header-footer.json',
```
Comment (anlinhs) [inline config/config.js:21]: Apparently we need to use the API management URLs, the one for int is: https://nhsuk-apim-int-uks.azure-api.net/menu I am told, if there is already a API subscription key in the app that can be used.

Comment (tomdoughty) [reply config/config.js:21]: I imagine the API subscription key in the application will be for a specific API product. This makes it quite difficult for anybody to pick this up and work on it as an open source project without having access to the subscription key

Comment (anlinhs) [reply config/config.js:21]: I don't know how it works but i was under the impression one subscription key could be used to access multiple endpoints? 

Where does open source come in to this?

Comment (tomdoughty) [reply config/config.js:21]: When I develop on these Node apps locally and want to view results I use my own personal "all products trial subscription" key from the API developer portal. I have never seen anything to suggest new endpoints are covered by that subscription.

In terms of open source I would expect a developer to be able to clone and run this application quickly and easily. However developers will now need to get hold of a key for it to run at all.

Comment (anlinhs) [reply config/config.js:21]: Ok i see, integration doesn't require a subscription key so you should have no problem running locally using that.

Comment (tomdoughty) [reply config/config.js:21]: Ahhh cool thats perfect then... I will set the default to that and get the headers set up for subscription-key for other environments 👍

File: README.md [hunk 4]
```diff
@@ -60,6 +60,9 @@ environment.
 | `SEARCH_API_HOST`                  | Host name for the [NHS Developer API](https://developer.api.nhs.uk/)                                | api.nhs.uk                                          | Yes        |
 | `SEARCH_API_KEY`                   | `subscription-key` for the [NHS Developer API](https://developer.api.nhs.uk/)                       |                                                     | Yes        |
 | `SEARCH_API_VERSION`               | Version of the [NHS Developer API](https://developer.api.nhs.uk/)                                   | 1                                                   | Yes        |
+| `HEADER_FOOTER_API_ENDPOINT`       | JSON endpoint for NHS.UK header and footer API                                                      | `https://nhsuk-live.nhswebsite-integration.nhs.uk/content-api/header-footer.json` | Yes        |
```
Comment (anlinhs) [inline README.md:63]: Only minor but this needs updating to the correct URL too.

File: app/middleware/headerAndFooter.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        headerLinks: cachedData.headerLinks,
+      };
+      return next();
+    }
+
+    // Get data from Wagtail endpoint
+    const { header, footer } = await rp(endpoint, {
+      headers: {
+        'Content-Type': 'application/json',
+        'subscription-key': subscriptionKey,
```
Comment (anlinhs) [inline app/middleware/headerAndFooter.js:26]: The key for headers (I believe is case sensitive) so it should be Subscription-Key (capital S)

Comment (tomdoughty) [reply app/middleware/headerAndFooter.js]: Ahh we use lowercase on all our API documentation but I will change it to be consistent with Nuget package 👍

**Review by anlinhs — APPROVED** [review]

**Pull request merged**